src/distributed_copy.c

```diff
@@ -0,0 +1,857 @@
+
+	shardConnections->placements =
+		(PlacementConnection *) palloc0(sizeof(PlacementConnection) * placementCount);
+	placementCount = 0;
+
+	foreach(taskPlacementCell, finalizedPlacementList)
+	{
+		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+		char *nodeName = taskPlacement->nodeName;
+		PGconn *conn = NULL;
+		conn = ConnectToNode(nodeName, taskPlacement->nodePort);
+		if (conn != NULL)
+		{
+			char const *copy = ConstructCopyStatement(copyStatement, shardId);
+			shardConnections->placements[placementCount].conn = conn;
+			shardConnections->placements[placementCount].id = taskPlacement->id;
+			placementCount += 1;
+
+			/*
+			 * New connection: start transaction with copy command on it.
+			 * Append shard id to table name.
+			 */
+			if (!transactionManager->Begin(conn) ||
+				!PgShardExecute(conn, PGRES_COPY_IN, copy))
+			{
+				failedPlacementList = lappend(failedPlacementList, taskPlacement);
+				ereport(WARNING, (errcode(ERRCODE_IO_ERROR),
+								  errmsg("Failed to start '%s' on node %s:%d", 
+										 copy, nodeName, taskPlacement->nodePort)));
+			}
+		}
+		else
+		{
+			failedPlacementList = lappend(failedPlacementList, taskPlacement);
+			ereport(WARNING, (errcode(ERRCODE_IO_ERROR),
+							  errmsg("Failed to connect to node %s:%d", 
+									 nodeName, taskPlacement->nodePort)));
+		}
+	}
+
+	/* if all placements failed, error out */
+	if (list_length(failedPlacementList) == list_length(finalizedPlacementList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_IO_ERROR),
+						errmsg("Could not copy to any active placements for shard %ld",
+							   (long) shardId)));
+
+		/* otherwise, mark failed placements as inactive: they're stale */
+	}
+	foreach(failedPlacementCell, failedPlacementList)
+	{
+		ShardPlacement *failedPlacement = (ShardPlacement *) lfirst(failedPlacementCell);
+
+		UpdateShardPlacementRowState(failedPlacement->id, STATE_INACTIVE);
+	}
+	shardConnections->replicaCount = placementCount;
+}
+
+/*
+ * Create single-linked list from hash table entries
+ */
+static List *
+HTABToList(HTAB *hash)
+{
+	HASH_SEQ_STATUS hashCursor;
+	void *entry;
+	List *list = NULL;
+
+	hash_seq_init(&hashCursor, hash);
+	while ((entry = hash_seq_search(&hashCursor)) != NULL)
+	{
+		list = lappend(list, entry);
+	}
+	return list;
+}
+
+/*
+ * Copy data from sepcified table
+ */
+static void
+PgShardCopyTo(CopyStmt *copyStatement, char const *query)
+{
+	RangeVar *relation = copyStatement->relation;
+	uint64 processedCount = 0;
+	char const *qualifiedName = quote_qualified_identifier(relation->schemaname,
+														   relation->relname);
+	List *queryList = raw_parser(psprintf("select * from %s", qualifiedName));
+	copyStatement->query = linitial(queryList);
+	copyStatement->relation = NULL;
+	
+	/* Collecting data from shards will be done by select handler */
+	DoCopy(copyStatement, query, &processedCount);
+}	
+
+static int
+CompareShardIntervalsByMinHashToken(const void *leftElement, 
```

Comment:
The name of the function contains HashToken, but this should be Value since the value may not be a hash token.

s/CompareShardIntervalsByMinHashToken/CompareShardIntervalsByMinValue/